src/Voron/Impl/PagerState.cs

```diff
@@ -100,24 +100,75 @@ public PagerState(AbstractPager pager, long prefetchSegmentSize, long prefetchRe
 
         public readonly byte* MapBase; // { get; set; }
 
-        internal bool _released;
-
         public void Release()
         {
-            if (Interlocked.Decrement(ref _refs) != 0)
+            int currentRefCount;
+            int newRefCount;
+
+            // Uses atomic compare-and-swap to safely decrement the reference count,
+            // ensuring consistency even with concurrent modifications by other threads.
+            do
+            {
+                currentRefCount = _refs;
+
+                // Checks for negative reference counts to detect potential misuse or logical errors,
+                // such as double releases or illegal state transitions, preventing further incorrect decrements.
+                if (currentRefCount < 0)
```

Comment:
Here you'll allow it to be 0, then drop to -1, no?

Reply:
Shouldn't it throw an error in state 0 too? That means we tried releasing more times than we have references.

Reply:
Yes, that's correct. Updated.